framework/src/meshgenerators/SubdomainElementIDs.C

```diff
@@ -0,0 +1,110 @@
+                                                      "Subdomain names present in the input mesh");
+  params.addRequiredParam<std::vector<std::string>>("element_id_names",
+                                                    "List of user-defined element ID names");
+  params.addParam<std::vector<std::vector<dof_id_type>>>(
+      "element_ids", "User-defined element IDs corresponding to 'subdomains' in the same order");
+
+  params.addParam<std::vector<dof_id_type>>("default_element_ids",
+                                            "Default element IDs for elements not in 'subdomains'");
+
+  params.addClassDescription("Assign element IDs for elements on a mesh based on mesh subdomains.");
+  return params;
+}
+
+SubdomainElementIDs::SubdomainElementIDs(const InputParameters & params)
+  : MeshGenerator(params),
+    _input(getMesh("input")),
+    _subdomain_names(getParam<std::vector<SubdomainName>>("subdomains"))
+{
+  if (_subdomain_names.size() == 0)
+    paramError("subdomains", "Empty subdomain vector provided!");
+}
+
+std::unique_ptr<MeshBase>
+SubdomainElementIDs::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  // construct a map from the subdomain ID to the index in 'subdomain'
+  std::map<SubdomainID, unsigned int> subdomains;
+  for (unsigned int i = 0; i < _subdomain_names.size(); ++i)
+  {
+    SubdomainID id = Moose::INVALID_BLOCK_ID;
+    std::istringstream ss(_subdomain_names[i]);
+
+    if (!(ss >> id) || !ss.eof())
+      id = mesh->get_id_by_name(_subdomain_names[i]);
+
+    auto check = subdomains.insert(std::make_pair(id, i));
+    if (!check.second)
+      paramError("subdomains", "Cannot have subdomain with ID ", id, " listed more than once!");
+  }
+
+  auto & element_id_names = getParam<std::vector<std::string>>("element_id_names");
+  auto & element_ids = getParam<std::vector<std::vector<dof_id_type>>>("element_ids");
+
+  if (element_id_names.size() != element_ids.size())
+    paramError("element_ids", "Inconsistent vector size for elmenet IDs");
+  for (auto & element_id : element_ids)
+  {
+    if (_subdomain_names.size() != element_id.size())
+      paramError("element_ids", "Inconsistent vector size for element IDs");
+  }
+
+  std::vector<unsigned int> xid;
+  for (auto & element_id_name : element_id_names)
+  {
+    if (!mesh->has_elem_integer(element_id_name))
+      xid.push_back(mesh->add_elem_integer(element_id_name));
+    else
+      xid.push_back(mesh->get_elem_integer_index(element_id_name));
+  }
+
+  if (isParamValid("default_element_ids"))
+  {
+    auto & default_ids = getParam<std::vector<dof_id_type>>("default_element_ids");
+    if (default_ids.size() != element_id_names.size())
+      paramError("default_element_ids", "Inconsistent vector size for default element IDs");
+
+    for (auto & elem : mesh->element_ptr_range())
+    {
+      for (unsigned int i = 0; i < element_ids.size(); ++i)
+        elem->set_extra_integer(xid[i], default_ids[i]);
+    }
+  }
```

Comment:
this seems inefficient. Setting the default everywhere to override it later.
Seems like an else statement in the next loop is cleaner

Reply:
There is a hidden logic here: when `default_element_ids` is not set, the elements whose subdomain IDs are not in the list will not get element IDs assigned or changed. It is difficult to accommodate this logic in one element loop. Not that looping though element is extremely fast, no need to worry efficiency unless there are some heavy operations like element initialization.